pkg/proc/bininfo.go

```diff
@@ -427,44 +435,128 @@ func (bi *BinaryInfo) PCToFunc(pc uint64) *Function {
+
+	dwarf       *dwarf.Data
+	dwarfReader *dwarf.Reader
+	loclist     loclistReader
+
+	typeCache map[dwarf.Offset]godwarf.Type
+
+	// runtimeTypeToDIE maps between the offset of a runtime._type in
+	// runtime.moduledata.types and the offset of the DIE in debug_info. This
+	// map is filled by using the extended attribute godwarf.AttrGoRuntimeType
+	// which was added in go 1.11.
+	runtimeTypeToDIE map[uint64]runtimeTypeDIE
+
+	loadErrMu sync.Mutex
+	loadErr   error
 }
 
-// AddImage adds the specified image to bi.
-func (bi *BinaryInfo) AddImage(path string, addr uint64) {
-	if !strings.HasPrefix(path, "/") {
-		return
+// hasImage returns true if the shared object has already been added.
+func (bi *BinaryInfo) hasImage(path string, addr uint64) bool {
+	if len(bi.Images) > 0 && !strings.HasPrefix(path, "/") {
+		return true
 	}
 	for _, image := range bi.Images {
 		if image.Path == path && image.addr == addr {
-			return
+			return true
 		}
 	}
-	//TODO(aarzilli): actually load informations about the image here
-	bi.Images = append(bi.Images, &Image{Path: path, addr: addr})
+	return false
+}
+
+// AddImage adds the specified image to bi, loading data asynchronously.
+// Addr is the relocated entry point for the executable and staticBase (i.e.
+// the relocation offset) for all other images.
+// The first image added must be the executable file.
+func (bi *BinaryInfo) AddImage(path string, addr uint64) error {
+	if bi.hasImage(path, addr) {
+		return nil
+	}
+	image := &Image{Path: path, addr: addr, typeCache: make(map[dwarf.Offset]godwarf.Type)}
+	// add Image regardless of error so that we don't attempt to re-add it every time we stop
+	image.index = len(bi.Images)
+	bi.Images = append(bi.Images, image)
+	err := loadBinaryInfo(bi, image, path, addr)
+	if err != nil {
+		bi.Images[len(bi.Images)-1].loadErr = err
+	}
+	return err
+}
+
+// moduleDataToImage finds the image corresponding to the given module data object.
+func (bi *BinaryInfo) moduleDataToImage(md *moduleData) *Image {
+	return bi.funcToImage(bi.PCToFunc(uint64(md.text)))
+}
+
+// imageToModuleData finds the module data in mds corresponding to the given image.
+func (bi *BinaryInfo) imageToModuleData(image *Image, mds []moduleData) *moduleData {
+	for _, md := range mds {
+		im2 := bi.moduleDataToImage(&md)
+		if im2.index == image.index {
+			return &md
+		}
+	}
+	return nil
+}
+
+// typeToImage returns the image containing the give type.
+func (bi *BinaryInfo) typeToImage(typ godwarf.Type) *Image {
+	return bi.Images[typ.Common().Index]
 }
 
 // Close closes all internal readers.
 func (bi *BinaryInfo) Close() error {
-	if bi.sepDebugCloser != nil {
-		bi.sepDebugCloser.Close()
+	var errs []error
+	for _, image := range bi.Images {
+		if image.sepDebugCloser != nil {
+			err := image.sepDebugCloser.Close()
+			if err != nil {
+				errs = append(errs, fmt.Errorf("closing shared object %q (split dwarf): %v", image.Path, err))
+			}
+		}
+		if image.closer != nil {
```

Comment:
nit: A refactor could possibly be to embed the closer in image so we can just call `image.Close()`.